Compile error in corner.cpp: ‘_mm256_setr_m128’ was not declared in this scope

##### System information (version)
- OpenCV => git master (currently at 3.2.0.r103.g74defefd6)
- Operating System / Platform => Arch Linux 64-bit
- Compiler => gcc 6.3.1

##### Detailed description
I'm getting the following compile error when trying to build OpenCV version git master:
```
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/imgproc/src/corner.cpp: In function ‘void cv::store_interleave(float*, const __m256&, const __m256&, const __m256&)’:
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/imgproc/src/corner.cpp:116:86: error: ‘_mm256_setr_m128’ was not declared in this scope
     _mm256_storeu_ps(ptr, _mm256_setr_m128(_mm_castsi128_ps(a0), _mm_castsi128_ps(b0)));
                                                                                      ^
```

This issue seems to be caused by the recent git commit aa5caf83f6443a4581b1d52d23d581e7fee6f919.

##### Steps to reproduce
Attached.
Complete output also provided.

[opencv-steps.txt](https://github.com/opencv/opencv/files/732465/opencv-steps.txt)
[opencv-output.txt](https://github.com/opencv/opencv/files/732466/opencv-output.txt)

--- a/modules/imgproc/src/corner.cpp
+++ b/modules/imgproc/src/corner.cpp
@@ -113,9 +113,16 @@ static inline void store_interleave(float* ptr, const __m256& a, const __m256& b
     v_transpose4x4(v_uint32x4(a1), v_uint32x4(b1), v_uint32x4(c1), z, u0, u1, u2, u3);
     v_pack4x3to3x4(u0.val, u1.val, u2.val, u3.val, a1, b1, c1);
 
+#if !defined(__GNUC__) || defined(__INTEL_COMPILER)
     _mm256_storeu_ps(ptr, _mm256_setr_m128(_mm_castsi128_ps(a0), _mm_castsi128_ps(b0)));
     _mm256_storeu_ps(ptr + 8, _mm256_setr_m128(_mm_castsi128_ps(c0), _mm_castsi128_ps(a1)));
     _mm256_storeu_ps(ptr + 16,  _mm256_setr_m128(_mm_castsi128_ps(b1), _mm_castsi128_ps(c1)));
+#else
+    // GCC: workaround for missing AVX intrinsic: "_mm256_setr_m128()"
+    _mm256_storeu_ps(ptr, _mm256_insertf128_ps(_mm256_castps128_ps256(_mm_castsi128_ps(a0)), _mm_castsi128_ps(b0), 1));
+    _mm256_storeu_ps(ptr + 8, _mm256_insertf128_ps(_mm256_castps128_ps256(_mm_castsi128_ps(c0)), _mm_castsi128_ps(a1), 1));
+    _mm256_storeu_ps(ptr + 16,  _mm256_insertf128_ps(_mm256_castps128_ps256(_mm_castsi128_ps(b1)), _mm_castsi128_ps(c1), 1));
+#endif
 }
 #endif // CV_AVX
 

